tests/integration/e2e/conftest.py

```diff
@@ -162,23 +163,106 @@ async def _deploy_client(role: Literal["producer", "consumer"]):
             num_units=1,
             series="jammy",
             channel="edge",
-            config={"topic-name": "HOT-TOPIC", "extra-user-roles": role},
+            config=config,
         )
         await ops_test.model.wait_for_idle(apps=[generated_app_name])
 
+        return generated_app_name
+
+    logger.info(f"setting up data_integrator - current apps {apps}")
+    yield _deploy_data_integrator
+
+    logger.info(f"tearing down {apps}")
+    for app in apps:
+        logger.info(f"tearing down {app}")
+        await ops_test.model.applications[app].remove()
+
+    await ops_test.model.wait_for_idle(apps=[kafka], idle_period=30, status="active", timeout=1800)
+
+
+@pytest.fixture(scope="function")
+async def deploy_test_app(ops_test: OpsTest, kafka, tls):
+    """Factory fixture for deploying + tearing down client applications."""
+    # tracks deployed app names for teardown later
+    apps = []
+
+    async def _deploy_test_app(
+        role: Literal["producer", "consumer"],
+        topic_name: str = "test-topic",
+        consumer_group_prefix: Optional[str] = None,
+        num_messages: int = 1500,
```

Comment:
**todo** best to use 2000. In the password rotation tests, 1500 is a bit edgy, and we risk that when the second consumers spins the 1500 messages have already been processed